plugin/trino-hive/src/main/java/io/trino/plugin/hive/s3/TrinoS3FileSystem.java

```diff
@@ -1572,17 +1574,11 @@ public void close()
         private void flushBuffer(boolean finished)
                 throws IOException
         {
-            try {
-                waitForPreviousUploadFinish();
-            }
-            catch (IOException e) {
-                failed = true;
-                abortUploadSuppressed(e);
-                throw e;
-            }
+            // Skip multipart upload if there would only be one part
 
-            // skip multipart upload if there would only be one part
-            if (finished && uploadId.isEmpty()) {
+            // We use multipartUploadStarted flag here instead  uploadId.ifPresent() because visibility of latter one is not
+            // ensured here. It is mutated in other thread spawned by uploadExecutor.submit below
```

Comment:
This comment is understandable from reviewer and historical perspective. 
We wouldn't write such comment if we had the code right from the start, which suggests the comment is more suitable for the commit message than the code.